Fixed issue with deleted containers and publish:

abasic opened the pull request:
- After publishing, published files of deleted containers are removed
- After unpublishing, files are bulk deleted instead of one at a time
- Added deleteFiles method to storage providers
- Updated listFiles methods for storage providers to return the same data structure

Closes #400

abasic committed:
Fixed issue with deleted containers and publish:

- now the activity folder is cleaned before publishing

abasic committed:
Reworked unpublishing of deleted content:

- removed clearing of the folder before publlsh
- added removing of redundant files after publishing
- updated storage providers with deleteFiles method
- removed cleanFolder method from storage providers
- update listFiles method of storage providers
- updated activity unpublishing so that it now bulk deletes files

kronicker commented on server/shared/publishing/helpers.js:
```diff
@@ -254,8 +278,8 @@ function defaultSummaryBuilder({ id, uid, type, publishedAs, elements = [] }) {
 
-function getContentContainerFilenames({ contentContainers = [] }) {
-  return map(contentContainers, it => `${it.id}.${it.publishedAs}`);
+function getContainerFilePaths(baseUrl, containers = []) {
```
`getContainersFilePaths` since its array of containers?

kronicker commented on server/shared/publishing/helpers.js:
```diff
@@ -89,19 +95,21 @@ function unpublishActivity(repository, activity) {
+      .map(deleted, it => {
+        const baseUrl = getBaseUrl(repository.id, it.id);
+        const filePaths = getContainerFilePaths(baseUrl, it.contentContainers);
+
```
I would say that this space is not necessary since the flow between the lines is consistent and obvious

kronicker commented on server/shared/publishing/helpers.js:
```diff
@@ -183,6 +195,18 @@ async function fetchCustomContainers(parent, config) {
+  const redundantFilePaths = difference(publishedFilePaths, filePaths);
+  if (redundantFilePaths.length) await storage.deleteFiles(redundantFilePaths);
+
+  return containers;
```
I think you would benefit in readability from the promise chain here rather than async/await with blank spaces.

kronicker replied:
Also, the signature of this method is a bit misleading for me since it's named `unpublishDeletedContainers` while it receives published containers that are to be preserved and is returning that original set of containers as a result 🤔

abasic replied:
> I think you would benefit in readability from the promise chain here rather than async/await with blank spaces.

Did you mean something like this:
```js
function unpublishDeletedContainers(parent, containers) {
  const baseUrl = getBaseUrl(parent.repositoryId, parent.id);
  const filePaths = getContainersFilePaths(baseUrl, containers);
  return storage
    .listFiles(baseUrl)
    .then(publishedFilePaths => {
      const redundantFilePaths = difference(publishedFilePaths, filePaths);
      if (redundantFilePaths.length) return storage.deleteFiles(redundantFilePaths);
    });
}
```

underscope replied:
@kronicker ☝️ 👀

kronicker replied:
@abasic 👍

abasic committed:
Refactored publish helpers:

- renamed getContainerFilePaths to getContainersFilePaths
- refactored unpublishDeletedContainers to use promise chain

underscope commented on server/shared/publishing/helpers.js:
```diff
@@ -89,19 +95,20 @@ function unpublishActivity(repository, activity) {
+      .map(deleted, it => {
+        const baseUrl = getBaseUrl(repository.id, it.id);
+        const filePaths = getContainersFilePaths(baseUrl, it.contentContainers);
+        return storage.deleteFiles(filePaths);
```
@abasic Since this is not transactional should we save spine first in case some of the containers fail or something breaks repo is not corrupt from consumer end)? 🤔

abasic replied:
If this is just a clean up and it is a nice to have that these are deleted then yes, we should move it after the entire activity is saved. We could even do a cron job or something of the sorts that makes sure everything is cleaned up. But, outside of this, at some point we should find a way to make the process of publishing rollback-able.
